xcvm/SPEC.md

```diff
@@ -364,4 +350,38 @@ Spawn XYZ BridgeSecurity::Deterministic 0 [
        Transfer USER { USDC: Ratio::ALL }       // On ABC, we transfer all USDC to the user. 
        ] { USDC: ALL },                         // We send over all our USDC back to ABC.
    ] { DOT: UNIT 100 },                         // We send over 100 DOT from ABC to XYZ.
-```
\ No newline at end of file
+```
+
+### Self-Referential Calls
+
+Many contracts have public functions with a `to` parameter, which indicates where funds should be sent to. Uniswap for example has the following interface for doing a [call](https://github.com/Uniswap/v2-core/blob/ee547b17853e71ed4e0101ccfd52e70d5acded58/contracts/interfaces/IUniswapV2Callee.sol#L4):
+
+```solidity
+function uniswapV2Call(address sender, uint amount0, uint amount1, bytes calldata data) external;
+```
+
+To properly construct this call and return funds to the interpreter instance, the `sender` parameter needs to be set to the interpreter's address. Normally this is an easy operation, as you can simply call the contract and pass `self`. However, from another chain, this is more difficult, as the address of the instance is unknown. There are three approaches to figuring out the interpreter's address:
+
+#### CREATE2
+
+The EVM has a specific [instruction](https://docs.openzeppelin.com/cli/2.8/deploying-with-create2#create2) which can be used to determine what the address of a contract will be based on keccak(0xFF, sender, salt, bytecode). In our case, the sender is the `XCVM Router` for that specific chain and known beforehand, the salt is the `Account, Salt` pair from the spawn call, and the bytecode is known beforehand as well. This means that when constructing the encoded uniswapV2Call, we can compute the `to` parameter on the origin chain. `CREATE2` is not available everywhere, however.
+
+#### Remote Instantiation First
```

Comment:
I seems that either chain support stuff (create2) or relayer does stuff, so may omit third way to complicate programs for end users.

Reply:
We can't have the relayer change data, as we use trustless proofs for data authenticity.

Reply:
can contracts have `self` or `this`(and `parent` may be:) ) as special keyword, so they are resolved as stuff executed? instead of direct `query` in xcvm program code.

Reply:
i am no deep in the topic, but if relayer can set some data permission less and then it can be validated on chain, than it can set data. so kind if can change data in certain ways, not sure if that fits here.

Reply:
solved this with late bindings.

Reply:
solved this with late bindings.